contracts/src/v0.8/dev/VRFSubscriptionBalanceMonitor.sol

```diff
@@ -0,0 +1,314 @@
+// SPDX-License-Identifier: MIT
+
+pragma solidity 0.8.6;
+
+import "../ConfirmedOwner.sol";
+import "../interfaces/KeeperCompatibleInterface.sol";
+import "../interfaces/VRFCoordinatorV2Interface.sol";
+import "../interfaces/LinkTokenInterface.sol";
+import "@openzeppelin/contracts/security/Pausable.sol";
+
+/**
+ * @title The VRFSubscriptionBalanceMonitor contract.
+ * @notice A keeper-compatible contract that monitors and funds VRF subscriptions.
+ */
+contract VRFSubscriptionBalanceMonitor is ConfirmedOwner, Pausable, KeeperCompatibleInterface {
+  VRFCoordinatorV2Interface public COORDINATOR;
+  LinkTokenInterface public LINKTOKEN;
+
+  uint256 private constant MIN_GAS_FOR_TRANSFER = 55_000;
+
+  event FundsAdded(uint256 amountAdded, uint256 newBalance, address sender);
+  event FundsWithdrawn(uint256 amountWithdrawn, address payee);
+  event TopUpSucceeded(uint64 indexed subscriptionId);
+  event TopUpFailed(uint64 indexed subscriptionId);
+  event KeeperRegistryAddressUpdated(address oldAddress, address newAddress);
+  event VRFCoordinatorV2AddressUpdated(address oldAddress, address newAddress);
+  event LinkTokenAddressUpdated(address oldAddress, address newAddress);
+  event MinWaitPeriodUpdated(uint256 oldMinWaitPeriod, uint256 newMinWaitPeriod);
+
+  error InvalidWatchList();
+  error OnlyKeeperRegistry();
+  error DuplicateSubcriptionId(uint64 duplicate);
+
+  struct Target {
+    bool isActive;
+    uint96 minBalanceJuels;
+    uint96 topUpAmountJuels;
+    uint56 lastTopUpTimestamp;
+  }
+
+  address private s_keeperRegistryAddress;
+  uint256 private s_minWaitPeriodSeconds;
+  uint64[] private s_watchList;
+  mapping(uint64 => Target) internal s_targets;
+
+  /**
+   * @param linkTokenAddress the Link token address
+   * @param coordinatorAddress the address of the vrf coordinator contract
+   * @param keeperRegistryAddress the address of the keeper registry contract
+   * @param minWaitPeriodSeconds the minimum wait period for addresses between funding
+   */
+  constructor(
+    address linkTokenAddress,
+    address coordinatorAddress,
+    address keeperRegistryAddress,
+    uint256 minWaitPeriodSeconds
+  ) ConfirmedOwner(msg.sender) {
+    setLinkTokenAddress(linkTokenAddress);
+    setVRFCoordinatorV2Address(coordinatorAddress);
+    setKeeperRegistryAddress(keeperRegistryAddress);
+    setMinWaitPeriodSeconds(minWaitPeriodSeconds);
+  }
+
+  /**
+   * @notice Sets the list of subscriptions to watch and their funding parameters.
+   * @param subscriptionIds the list of subscription ids to watch
+   * @param minBalancesJuels the minimum balances for each subscription
+   * @param topUpAmountsJuels the amount to top up each subscription
+   */
+  function setWatchList(
+    uint64[] calldata subscriptionIds,
+    uint96[] calldata minBalancesJuels,
+    uint96[] calldata topUpAmountsJuels
+  ) external onlyOwner {
```

Comment:
I feel like this might be simpler if this simply does a pure overwrite - I feel like I'd have to read this function code everytime to figure out what it does on successive calls because it depends on the state of `s_targets`. Simpler to just overwrite everytime and is easier to understand for people interacting with the contract.